include/deal.II/distributed/fullydistributed_tria.h

```diff
@@ -0,0 +1,237 @@
+        cell_iterator;
+
+      typedef
+        typename dealii::Triangulation<dim, spacedim>::active_cell_iterator
+          active_cell_iterator;
+
+      typedef
+        typename dealii::Triangulation<dim, spacedim>::CellStatus CellStatus;
+
+      void
+      reinit(ConstructionData<dim, spacedim> &construction_data);
+
+      virtual void
+      create_triangulation(const std::vector<Point<spacedim>> &vertices,
+                           const std::vector<CellData<dim>> &  cells,
+                           const SubCellData &subcelldata) override;
+
+
+      enum Settings
+      {
+        /**
+         * Default settings, other options are disabled.
+         */
+        default_setting = 0x0,
+        /**
+         * This flags needs to be set to use the geometric multigrid
+         * functionality. This option requires additional computation and
+         * communication. Note: geometric multigrid is still a work in
+         * progress.
+         */
+        construct_multigrid_hierarchy = 0x1
+      };
+
+
+      Triangulation(MPI_Comm mpi_communicator,
+                    Settings settings_ = default_setting);
+
+      Triangulation(MPI_Comm mpi_communicator,
+                    MPI_Comm mpi_communicator_coarse,
+                    Settings settings_ = default_setting);
+
+      virtual ~Triangulation() = default;
+
+      virtual void
+      execute_coarsening_and_refinement() override;
+
+      virtual bool
+      prepare_coarsening_and_refinement() override;
+
+      virtual bool
+      has_hanging_nodes() const override;
+
+      virtual std::size_t
+      memory_consumption() const override;
+
+      virtual void
+      add_periodicity(
+        const std::vector<GridTools::PeriodicFacePair<cell_iterator>> &)
+        override;
+
+      virtual void
+      update_number_cache() override;
+
+      virtual std::map<unsigned int, std::set<dealii::types::subdomain_id>>
+      compute_vertices_with_ghost_neighbors() const override;
+
+      bool
+      is_multilevel_hierarchy_constructed() const override;
+
+      MPI_Comm
+      get_coarse_communicator() const;
+
+      virtual unsigned int
+      coarse_cell_id_to_coarse_cell_index(
+        const types::coarse_cell_id coarse_cell_id) const override;
+
+      virtual types::coarse_cell_id
+      coarse_cell_index_to_coarse_cell_id(
+        const unsigned int coarse_cell_index) const override;
+
+    private:
+      /**
+       * store the Settings.
+       */
+      Settings settings;
+
+      template <typename>
+      friend class dealii::internal::DoFHandlerImplementation::Policy::
+        ParallelDistributed;
+
+      template <int, int, class>
+      friend class dealii::FETools::internal::ExtrapolateImplementation;
+
+      std::vector<std::pair<types::coarse_cell_id, unsigned int>>
+        coarse_gid_to_lid;
+      std::vector<std::pair<unsigned int, types::coarse_cell_id>>
+        coarse_lid_to_gid;
```

Comment:
I would have expected that it is enough to store a std::vector< types::coarse_cell_id> of length n_cells(0) here. Can this have gaps?

Reply:
You are right having a simple `std::vector< types::coarse_cell_id>` is better. I have changed it.